parsec/core/local_device.py

```diff
@@ -0,0 +1,242 @@
+from parsec.core.types import EntryID, LocalDevice, BackendOrganizationAddr
+
+
+class LocalDeviceError(Exception):
+    pass
+
+
+class LocalDeviceCryptoError(LocalDeviceError):
+    pass
+
+
+class LocalDeviceNotFoundError(LocalDeviceError):
+    pass
+
+
+class LocalDeviceAlreadyExistsError(LocalDeviceError):
+    pass
+
+
+class LocalDeviceValidationError(LocalDeviceError):
+    pass
+
+
+class LocalDevicePackingError(LocalDeviceError):
+    pass
+
+
+class DeviceFileSchema(BaseSchema):
+    type = fields.CheckedConstant("password", required=True)
+    salt = fields.Bytes(required=True)
+    ciphertext = fields.Bytes(required=True)
+
+    # Since human_handle/device_label has been introduced, device_id is
+    # redacted (i.e. user_id and device_name are 2 random uuids), hence
+    # those fields have been added to the device file so the login page in
+    # the GUI can use them to provide useful information.
+    human_handle = HumanHandleField(allow_none=True, missing=None)
+    device_label = fields.String(allow_none=True, missing=None)
+
+
+key_file_serializer = MsgpackSerializer(
+    DeviceFileSchema, validation_exc=LocalDeviceValidationError, packing_exc=LocalDevicePackingError
+)
+
+
+def generate_new_device(
+    device_id: DeviceID,
+    organization_addr: BackendOrganizationAddr,
+    profile: UserProfile = UserProfile.STANDARD,
+    human_handle: Optional[HumanHandle] = None,
+    device_label: Optional[str] = None,
+    signing_key: Optional[SigningKey] = None,
+    private_key: Optional[PrivateKey] = None,
+) -> LocalDevice:
+    return LocalDevice(
+        organization_addr=organization_addr,
+        device_id=device_id,
+        device_label=device_label,
+        human_handle=human_handle,
+        signing_key=signing_key or SigningKey.generate(),
+        private_key=private_key or PrivateKey.generate(),
+        profile=profile,
+        user_manifest_id=EntryID(),
+        user_manifest_key=SecretKey.generate(),
+        local_symkey=SecretKey.generate(),
+    )
+
+
+def get_key_file(config_dir: Path, device: LocalDevice) -> Path:
+    slug = device.slug
+    return get_devices_dir(config_dir) / slug / f"{slug}.keys"
+
+
+def get_devices_dir(config_dir: Path) -> Path:
+    return config_dir / "devices"
+
+
+@attr.s(slots=True, frozen=True, auto_attribs=True)
+class AvailableDevice:
+    key_file_path: Path
+    organization_id: OrganizationID
+    device_id: DeviceID
+    human_handle: Optional[HumanHandle]
+    device_label: Optional[str]
+
+    @property
+    def user_display(self) -> str:
+        return str(self.human_handle or self.device_id.user_id)
+
+    @property
+    def device_display(self) -> str:
+        return self.device_label or str(self.device_id.device_name)
+
+    @property
+    def slug(self) -> str:
+        # Drop the `.keys` suffix
+        return self.key_file_path.name.rsplit(".", 1)[0]
+
+    @property
+    def slughash(self) -> str:
+        return sha256(self.slug.encode()).hexdigest()
```

Comment:
Necessary code duplication?
https://github.com/Scille/parsec-cloud/pull/1174/commits/6e39a87268553f364a2564fe8545e882654254b1#diff-310fbdc384848dedf0feca35dfb5f58cR80-R88

Reply:
Yep, the `LocalDevice` and `AvailableDevice` both implement this.
But given it's really trivial stuff I find it better than introducing a shared 1-liner function